File: clients/python/examples/pty.py
#!/usr/bin/env python3
"""Example: spawn a PTY session and relay I/O."""

import asyncio
import sys
from bashclient import BashClient


async def main():
    socket_path = sys.argv[1] if len(sys.argv) > 1 else "/tmp/bash-server/sock"
    token = sys.argv[2] if len(sys.argv) > 2 else "a" * 64

    async with await BashClient.connect(socket_path) as client:
        await client.auth(token)

        exit_event = asyncio.Event()

        def on_output(data: str):
            sys.stdout.write(data)
            sys.stdout.flush()

        def on_exit(code: int):
            print(f"\n[PTY exited with code {code}]")
            exit_event.set()

        client.pty.on("output", on_output)
        client.pty.on("exit", on_exit)

        # Spawn PTY
        info = await client.pty.spawn(rows=24, cols=80, strip_ansi=True)
        print(f"PTY spawned: pid={info.pid} {info.rows}x{info.cols}")

        # Send a command
        await client.pty.write_input("echo 'Hello from PTY!'\n")
        await asyncio.sleep(0.5)

        # Send exit
        await client.pty.write_input("exit\n")

        # Wait for exit
        try:
            await asyncio.wait_for(exit_event.wait(), timeout=5.0)
        except asyncio.TimeoutError:
            print("[Timeout waiting for PTY exit]")
            await client.pty.close()


if __name__ == "__main__":
    asyncio.run(main())

File: clients/typescript/examples/pty.ts
import { BashClient } from "../src";

async function main() {
  const socketPath = process.argv[2] ?? "/tmp/bash-server-1000/sock";
  const token = process.argv[3] ?? "a".repeat(64);

  const client = await BashClient.connect(socketPath);
  try {
    await client.auth(token);

    let exited = false;

    client.pty.on("output", (data: string) => {
      process.stdout.write(data);
    });
    client.pty.on("exit", (code: number) => {
      console.log(`\n[PTY exited with code ${code}]`);
      exited = true;
    });

    const info = await client.pty.spawn({ rows: 24, cols: 80, strip_ansi: true });
    console.log(`PTY spawned: pid=${info.pid} ${info.rows}x${info.cols}`);

    await client.pty.writeInput("echo 'Hello from PTY!'\n");
    await new Promise((r) => setTimeout(r, 500));
    await client.pty.writeInput("exit\n");
    await new Promise((r) => setTimeout(r, 2000));

    if (!exited) await client.pty.close();
  } finally {
    await client.close();
  }
}

main().catch(console.error);
